crates/walrus-service/src/node.rs

```diff
@@ -2204,10 +2241,170 @@ impl StorageNode {
 }
 
 impl StorageNodeInner {
+    /// Computes a deferral duration from the configured policy and a given unencoded size.
+    pub fn deferral_for_unencoded_size(&self, size_bytes: u64) -> Option<std::time::Duration> {
+        self.live_upload_deferral_config
+            .deferral_for_size(size_bytes)
+    }
+
+    /// Returns the latest checkpoint sequence number known to the node.
+    pub fn latest_checkpoint_sequence_number(&self) -> Option<u64> {
+        self.event_manager.latest_checkpoint_sequence_number()
+    }
+
+    /// Applies a recovery deferral if the node determines the blob is part of an in-flight upload.
+    pub async fn maybe_apply_live_upload_deferral(
+        &self,
+        blob_id: BlobId,
+        checkpoint_sequence: u64,
+    ) {
+        if !self.live_upload_deferral_config.enabled {
+            return;
+        }
+
+        let Some(latest_checkpoint) = self.latest_checkpoint_sequence_number() else {
+            tracing::trace!(
+                %blob_id,
+                "latest checkpoint unknown; skipping live-upload deferral"
+            );
+            return;
+        };
+
+        if checkpoint_sequence > latest_checkpoint {
+            tracing::trace!(
+                %blob_id,
+                checkpoint_sequence,
+                latest_checkpoint,
+                "checkpoint sequence ahead of latest; skipping live-upload deferral"
+            );
+            return;
+        }
+
+        let lag = latest_checkpoint.saturating_sub(checkpoint_sequence);
+        if lag > self.live_upload_deferral_config.max_checkpoint_lag {
+            tracing::trace!(
+                %blob_id,
+                lag,
+                threshold = self.live_upload_deferral_config.max_checkpoint_lag,
+                "checkpoint lag too high; skipping live-upload deferral"
+            );
+            return;
+        }
+
+        let metadata = match self.storage.get_metadata(&blob_id) {
```

Comment:
Remark: It's a bit unfortunate that we don't have the unencoded size in the blob info...

Suggestion: Given that we only need this to compute the deferral duration, we could simply use a default value in case we don't have the metadata.